### PR TITLE
feat: Databricks streaming api

### PR DESCRIPTION
**Pull request by HenrikSommer**
## Description

With this pull request, we add the capability of streaming the data returned by Databricks as an "external link". 

We do this with two purposes:
- to increase performance (less time to complete request)
- to decrease memory use and garbage collection (fewer heap memory allocations)

For now, the streaming functionality is added as a separate method, to allow package-clients to keep using the original functionality.

## Quality

- [x] Documentation is updated
- [x] Release notes are updated
- [x] Package version is updated
- [x] Public types and methods are documented
- [x] Tests are implemented and executed locally

### CODE REVIEW

File: source/Databricks/source/SqlStatementExecution.UnitTests/Internal/DatabricksSqlStatementClientTests.cs [hunk 1]
```diff
@@ -74,6 +74,31 @@ public DatabricksSqlStatementClientTests()
...
+            .Returns(SqlResponse.CreateAsFailed(statementId));
+        var sut = builder
+            .AddHttpClientResponse(_running)
+            .AddHttpClientResponse(_failed)
+            .UseParser(parserMock.Object)
+            .Build();
+
+        // Act and assert
+        await Assert.ThrowsAsync<DatabricksSqlException>(async () => await sut.
+            StreamAsync("some sql", new List<SqlStatementParameter>()).ToListAsync());
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.UnitTests/Internal/DatabricksSqlStatementClientTests.cs:99]: Consider using FluentAssertions and split "act" from "assert" like:

```suggestion
        // Act
        Func<Task> act = async () => await sut.ExecuteAsync("some sql", new List<SqlStatementParameter>()).ToListAsync();

        // Assert
        await act.Should().ThrowAsync<DatabricksSqlException>();
```

File: source/Databricks/source/SqlStatementExecution/Internal/DatabricksSqlStatementClient.cs [hunk 2]
```diff
@@ -166,16 +191,39 @@ private async Task<SqlChunkResponse> GetChunkAsync(string chunkLink)
...
-        return _responseResponseParser.ParseChunkDataResponse(jsonResponse, columnNames);
+        return httpResponse;
+    }
+
+    private void LogDatabricksRequest(string sqlStatement, List<SqlStatementParameter> sqlStatementParameters)
+    {
+        _logger.LogDebug(
+            "Executing SQL statement: {Sql}, with parameters: {Parameters}",
+            HttpUtility.HtmlEncode(sqlStatement),
+            sqlStatementParameters);
```
Comment (Sondergaard) [inline source/Databricks/source/SqlStatementExecution/Internal/DatabricksSqlStatementClient.cs:226]: Nit: Only HtmlEncode sqlStatement if log level debug is enabled

```suggestion
        if (_logger.IsEnabled(LogLevel.Debug))
        {
            _logger.LogDebug(
                "Executing SQL statement: {Sql}, with parameters: {Parameters}",
                HttpUtility.HtmlEncode(sqlStatement),
                sqlStatementParameters);
        }
```

File: source/Databricks/source/SqlStatementExecution/Internal/DatabricksSqlStatementClient.cs [hunk 3]
```diff
@@ -75,24 +73,50 @@ public class DatabricksSqlStatementClient : IDatabricksSqlStatementClient
...
+    }
+
+    public async IAsyncEnumerable<string[]> StreamAsync(
+        string sqlStatement,
+        List<SqlStatementParameter>? sqlStatementParameters = null)
+    {
+        sqlStatementParameters ??= new List<SqlStatementParameter>();
+        LogDatabricksRequest(sqlStatement, sqlStatementParameters);
+
+        var response = await GetFirstChunkOrNullAsync(sqlStatement, sqlStatementParameters).ConfigureAwait(false);
```
Comment (Sondergaard) [inline source/Databricks/source/SqlStatementExecution/Internal/DatabricksSqlStatementClient.cs:101]: The internals of GetFirstChunkOrNullAsync can be improved. It handles the response as a string and not a stream. The response is only passed on to another class that uses Newtonsoft. I would rather see the response being handled as a stream, and then use the performance System.Text.Json for parsing the stream.

I don't think this optimisation should be covered in this PR.

**Review by Sondergaard — APPROVED** [review]

File: source/Databricks/source/SqlStatementExecution.UnitTests/Internal/DatabricksSqlStatementClientTests.cs [hunk 4]
```diff
@@ -74,6 +74,31 @@ public DatabricksSqlStatementClientTests()
             ExecuteAsync("some sql", new List<SqlStatementParameter>()).ToListAsync());
     }
 
+    [Theory]
+    [InlineAutoMoqData]
+    public async Task StreamAsync_WhenFirstRunningThenStatementFails_ThrowsDatabricksSqlException(
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.UnitTests/Internal/DatabricksSqlStatementClientTests.cs:79]: Does this actually prove we fail in the expected scenario? It seems to me this exception is thrown for a number of reasons. :)

File: source/Databricks/source/SqlStatementExecution.UnitTests/Internal/SqlChunkDataResponseParserTests.cs [hunk 5]
```diff
@@ -33,24 +34,41 @@ public void Parse_ReturnsExpectedTableChunk(SqlChunkDataResponseParser sut)
...
+        public async Task ParseAsync_ReturnsExpectedTableChunk(SqlChunkDataResponseParser sut)
+        {
+            // Arrange
+            const string jsonResponse = "[[\"John\", \"Doe\"], [\"Jane\", \"Smith\"]]";
+
+            // Act
+            await using var jsonStream = new MemoryStream(Encoding.UTF8.GetBytes(jsonResponse));
+            var actual = sut.ParseAsync(jsonStream);
+
+            var actualList = await actual.ToListAsync();
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.UnitTests/Internal/SqlChunkDataResponseParserTests.cs:55]: // Assert?

File: source/Databricks/source/SqlStatementExecution.UnitTests/Internal/SqlChunkDataResponseParserTests.cs [hunk 6]
```diff
@@ -33,24 +34,41 @@ public void Parse_ReturnsExpectedTableChunk(SqlChunkDataResponseParser sut)
...
             string[] columnNames)
         {
             // Arrange
-            var jsonResponse = "invalid json";
+            const string jsonResponse = "invalid json";
 
             // Act & Assert
-            sut.Invoking(s => s.Parse(jsonResponse, columnNames))
-                .Should().Throw<Exception>();
+            sut.Invoking(s => s.Parse(jsonResponse, columnNames)).Should().Throw<Exception>();
```
Comment (dstenroejl) [inline source/Databricks/source/SqlStatementExecution.UnitTests/Internal/SqlChunkDataResponseParserTests.cs:71]: Can be split in act/assert, similar to what I showed in a previous comment. We should aim for 1 way of writing these kind of tests and be consistent. I'm not saying we have to refactor all tests, but when we add or refactor tests, we should follow 3xA (arrange, act, assert).

**Review by dstenroejl — COMMENTED** [review]
Added suggestions to how to write certain tests

**Review by dstenroejl — DISMISSED** [review]
LGTM

**Review by dstenroejl — APPROVED** [review]
LGTM

**Pull request merged**